cpp/src/arrow/gpu/cuda_context.h

```diff
@@ -37,13 +37,33 @@ class ARROW_EXPORT CudaDeviceManager {
   static Status GetInstance(CudaDeviceManager** manager);
 
   /// \brief Get the shared CUDA driver context for a particular device
+  /// \param[in] device_number
+  /// \param[out] out shared context
   Status GetContext(int gpu_number, std::shared_ptr<CudaContext>* ctx);
 
+  /// \brief Get the shared CUDA driver context for a particular device
+  /// \param[in] device_number
+  /// \param[in] handle CUDA context handler created by another library
+  /// \param[out] out shared context
+  Status GetSharedContext(int device_number, void* handle,
+                          std::shared_ptr<CudaContext>* out);
+
   /// \brief Create a new context for a given device number
+  /// \param[in] device_number
+  /// \param[out] out new context
   ///
   /// In general code will use GetContext
   Status CreateNewContext(int gpu_number, std::shared_ptr<CudaContext>* ctx);
 
+  /// \brief Create shared context for a given device number
```

Comment:
I admit I don't understand the difference between "get shared context" and "create shared context".

Reply:
"Create shared context" will create CudaContext with a context handle that is owned by another library (or created by another library).
"Get shared context" does "Create shared context" if the `CudaContext` for the given device is not created yet, and caches the `CudaContext` instance. Otherwise, it will return the cached instance.

This follows the same logic as in `GetContext` and `CreateNewContext` methods.

Reply:
Are all those variations actually useful? Isn't `GetSharedContext` enough?

Reply:
The difference between the values returned by `GetSharedContext` and `GetContext` is that in the latter case the created context is owned by arrow context manager (and that is allowed to destroy the context in `Close` method).
For the shared case, the context wrapped by `CudaContext` is assumed to be managed by another library.

Reply:
But why do we need both `GetSharedContext` and `CreateSharedContext`?